content/advent-2018/split-data-with-cdc.md

```diff
@@ -0,0 +1,232 @@
+
+```
+$ (echo -n foo; cat data) | go run main.go
+0       2908787 000c31839a100000        f41084d25bc273f2[...]
+2908787 2088949 001e8b4104900000        ba7c71226609c0de[...]
+4997736 1404824 000a18d1f0200000        9c4db21626a27d4b[...]
+6402560 617070  001bc6ac84300000        4c249c3b88500c23[...]
+7019630 1278326 001734984b000000        5bec562b7ad37b8b[...]
+```
+
+The SHA-256 hash for the first chunk changed from `8d17f5f7326a2fd7[...]` to `f41084d25bc273f2[...]`, but the rest is still the same. That's pretty cool.
+
+Let's see how many different chunks we have in our data file byte counting the different hashes:
+```
+$ cat data | go run main.go | cut -f4 | sort | uniq | wc -l
+69
+```
+
+What happens when we feed the program our data twice, how many chunks does it detect? Let's find out:
+
+```
+$ (cat data; cat data) | go run main.go | cut -f4 | sort | uniq | wc -l
+70
+```
+
+Huh, so we only got a single additional chunk. The cause is that for the first round, it'll find the same chunks as before (it's the same data after all) right until the last chunk. In the previous run, the end of the last chunk was determined by the end of the file. Now there's additional data, so the chunk continues. The next cut point will in the second run of the data, it'll be the same as the first cut point, just with some additional data at the beginning of the chunk, so the SHA-256 hash is different. Afterwards, the same chunks follow in the same order.
+
+Changing a few consecutive bytes of data somewhere in most cases also only affects a single chunk, let's write the output to a file, change the data a bit using `sed`, and write another file:
+
+```
+$ cat data | go run main.go > orig.log
+$ cat data | sed 's/EZE8HX/xxxxxx/' | go run main.go > mod.log
+```
+
+The string `EZE8HX` was present in my randomly generated file `data` only once and the `sed` command above changed it to the string `xxxxxx`. When we now compare the two log files using `diff`, we can see that for exactly one chunk the SHA-256 hash has changed, but the other chunks (and the chunk boundaries) stayed the same:
+
+```diff
+$ diff -au orig.log mod.log
+--- orig.log    2018-11-24 13:14:45.265329525 +0100
++++ mod.log     2018-11-24 13:19:32.344681407 +0100
+@@ -55,7 +55,7 @@
+ 83545247       547876  00198b8839f00000        2cbeea803ba79d54[...]
+ 84093123       889631  0007ddacabc00000        a2b6e8651ae1ab69[...]
+ 84982754       2561935 000c53712f500000        bca589959b019a80[...]
+-87544689       2744485 0000a49118b00000        02b125104f06ed85[...]
++87544689       2744485 0000a49118b00000        ce2e43496ae6e7c3[...]
+ 90289174       1167308 00034d6cce700000        ad25a331993d9db7[...]
+ 91456482       1719951 001d2fea8ae00000        ba5153845c5b5228[...]
+ 93176433       1362655 0003009fc1600000        8e7a35e340c10d61[...]
+```
+
+This technique can be used for many other things besides a backup program.
```

Comment:
It would be nice to have some examples here.

Reply:
I've added a few, good point!